### PR TITLE
Developer documentation - first part

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: docs/developers_guide/Components.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+- implementation
+
+Specification defines what are parameters of the component, and what are the results - i.e. how it transforms
+the Context. 
+                                 
+For simple components, the specification is derived from implementing method with the help of annotations. 
+
+## Specification
+                                                                                             
+### @MethodToInvoke
```
Comment (dswiecki) [inline docs/developers_guide/Components.md:16]: maybe lets give here links to our sample implementations

File: docs/developers_guide/Components.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+Other parameters allow the engine to construct the invocation (e.g. they represent the value of the kafka message), 
+so their value is not know when source/sink object (represented e.g. by Kafka consumer/producer) is created.
+The later kind of parameters is represented by `LazyParameter` trait. 
+Scenario author usually uses expressions containing `#input` variable etc. to define such parameter.
+During creation of component implementation
+their value is not known, but their exact type is known - as during scenario compilation the expression computing the value is compiled and typechecked.
+
+
+Please note that "eager" parameters - computed before creating implementation of the component - can also be represented with
```
Comment (dswiecki) [inline docs/developers_guide/Components.md:55]: I think I don't understand it :P

File: docs/developers_guide/Basics.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+- [ToJsonEncoder](https://github.com/TouK/nussknacker/blob/staging/utils/util/src/main/scala/pl/touk/nussknacker/engine/util/json/BestEffortJsonEncoder.scala)
+- [ObjectNamingProvider](https://github.com/TouK/nussknacker/blob/staging/utils/util/src/main/scala/pl/touk/nussknacker/engine/util/namespaces/ObjectNamingProvider.scala)
+- [ModelConfigLoader](https://github.com/TouK/nussknacker/blob/staging/interpreter/src/main/scala/pl/touk/nussknacker/engine/modelconfig/ModelConfigLoader.scala)
+- [ProcessMigrations](https://github.com/TouK/nussknacker/blob/staging/interpreter/src/main/scala/pl/touk/nussknacker/engine/migration/ProcessMigration.scala)
+- [DictServicesFactory](https://github.com/TouK/nussknacker/blob/staging/api/src/main/scala/pl/touk/nussknacker/engine/api/dict/DictServicesFactory.scala)
+
+### Designer customization
+
+- [ProcessChangeListenerFactory](https://github.com/TouK/nussknacker/blob/staging/ui/listener-api/src/main/scala/pl/touk/nussknacker/ui/listener/ProcessChangeListenerFactory.scala)
+- Security
```
[the remark this answers is not in the document]
Comment (mproch) [reply docs/developers_guide/Basics.md:53]: :+1: fixed indentation...

File: docs/developers_guide/Basics.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# Overview
+
+Please make sure you know common [Glossary](https://docs.nussknacker.io/documentation/about/GLOSSARY) and [SpEL](../scenarios_authoring/Spel) (especially the Data types section) before proceeding further.
```
[the remark this answers is not in the document]
Comment (mproch) [reply docs/developers_guide/Basics.md]: :+1: added .md

File: docs/developers_guide/Components.md (hunk 2 above)
Comment (mproch) [reply docs/developers_guide/Components.md:52]: I added example

File: docs/developers_guide/Components.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+This is the most powerful way to define components. It allows for:
+- Arbitrary ValidationContext transformations
+- Dynamic parameters. In particular:
+    - Parameters may depend on configuration. E.g., in OpenAPI enricher URL with definition is passed in configuration,
+      the parameters are generated on the base of this definition. 
+    - Parameters may depend on each other. Good example is Kafka/Schema registry Sink. The first parameter determines the target topic and its schema, which 
+      defines the rest of the parameters
+
+To
```
[the remark this answers is not in the document]
Comment (mproch) [reply docs/developers_guide/Components.md]: removed...

File: docs/developers_guide/Components.md (hunk 1 above)
Comment (mproch) [reply docs/developers_guide/Components.md:16]: I added a few links

File: docs/developers_guide/Basics.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mproch) [reply docs/developers_guide/Basics.md]: changed link to 'more relative'

**Review by dswiecki — APPROVED** [review]

**Pull request merged**